libclamav/cvd.c

```diff
@@ -519,197 +522,185 @@ void cl_cvdfree(struct cl_cvd *cvd)
-        return CL_EVERIFY;
-    }
-
-    free(md5);
-    cl_cvdfree(cvd);
-    return CL_SUCCESS;
+    return cl_cvdverify_ex(file, NULL);
 }
 
-cl_error_t cl_cvdverify(const char *file)
+cl_error_t cl_cvdverify_ex(const char *file, const char *certs_directory)
 {
     struct cl_engine *engine;
-    FILE *fs;
     cl_error_t ret;
-    int dbtype = 0;
-
-    if ((fs = fopen(file, "rb")) == NULL) {
-        cli_errmsg("cl_cvdverify: Can't open file %s\n", file);
-        return CL_EOPEN;
-    }
+    cvd_type dbtype = CVD_TYPE_CVD;
 
     if (!(engine = cl_engine_new())) {
         cli_errmsg("cl_cvdverify: Can't create new engine\n");
-        fclose(fs);
-        return CL_EMEM;
+        ret = CL_EMEM;
+        goto done;
     }
     engine->cb_stats_submit = NULL; /* Don't submit stats if we're just verifying a CVD */
 
-    if (!!cli_strbcasestr(file, ".cld"))
-        dbtype = 1;
-    else if (!!cli_strbcasestr(file, ".cud"))
-        dbtype = 2;
+    if (NULL != certs_directory) {
+        ret = cl_engine_set_str(engine, CL_ENGINE_CERTSDIR, certs_directory);
+        if (CL_SUCCESS != ret) {
+            cli_errmsg("cl_cvdverify: Failed to set engine certs directory\n");
+            goto done;
+        }
+    }
 
-    ret = cli_cvdload(fs, engine, NULL, CL_DB_STDOPT | CL_DB_PUA, dbtype, file, 1);
+    if (!!cli_strbcasestr(file, ".cld")) {
+        dbtype = CVD_TYPE_CLD;
+    } else if (!!cli_strbcasestr(file, ".cud")) {
+        dbtype = CVD_TYPE_CUD;
+    }
+
+    ret = cli_cvdload(engine, NULL, CL_DB_STDOPT | CL_DB_PUA, dbtype, file, 1);
 
+done:
     cl_engine_free(engine);
-    fclose(fs);
+
     return ret;
 }
 
-cl_error_t cli_cvdload(FILE *fs, struct cl_engine *engine, unsigned int *signo, unsigned int options, unsigned int dbtype, const char *filename, unsigned int chkonly)
+cl_error_t cli_cvdload(struct cl_engine *engine, unsigned int *signo, unsigned int options, cvd_type dbtype, const char *filename, unsigned int chkonly)
 {
-    struct cl_cvd cvd, dupcvd;
-    FILE *dupfs;
+    cl_error_t status = CL_ECVD;
     cl_error_t ret;
     time_t s_time;
-    int cfd;
     struct cli_dbio dbio;
-    struct cli_dbinfo *dbinfo = NULL;
-    char *dupname;
+    struct cli_dbinfo *dbinfo  = NULL;
+    char *dupname              = NULL;
+    cvd_t *cvd                 = NULL;
+    cvd_t *dupcvd              = NULL;
+    FFIError *cvd_open_error   = NULL;
+    FFIError *cvd_verify_error = NULL;
+    char *signer_name          = NULL;
 
     dbio.hashctx = NULL;
 
     cli_dbgmsg("in cli_cvdload()\n");
 
-    /* verify */
-    if ((ret = cli_cvdverify(fs, &cvd, dbtype)))
-        return ret;
+    /* Open the cvd and read the header */
+    cvd = cvd_open(filename, &cvd_open_error);
+    if (!cvd) {
+        cli_errmsg("cli_cvdload: Can't open CVD file %s: %s\n", filename, ffierror_fmt(cvd_open_error));
+        goto done;
+    }
+
+    /* For actual .cvd files, verify the digital signature. */
+    if (dbtype == CVD_TYPE_CVD) {
```

Comment:
Do we want to also check this for CUD?  Probably. This would be a case where database is not signed with old MD5-RSA, and only signed with external PKCS7 sig.